File: android/src/main/kotlin/ru/kardenvan/flutter_yandex_ads_plus/platform_api/model_factories/BannerAdViewArgumentsFactory.kt
package ru.kardenvan.flutter_yandex_ads_plus.platform_api.model_factories

import android.util.Size
import ru.kardenvan.flutter_yandex_ads_plus.ads.AdParameters
import ru.kardenvan.flutter_yandex_ads_plus.ads.banner_yandex_ad.BannerYandexAdViewArguments

class BannerAdViewArgumentsFactory {
    companion object Factory {
        fun fromMap(map: Map<*, *>): BannerYandexAdViewArguments {
            val uid = map["uid"]

            if (uid !is String) {
                throw Exception(
                    "Arguments map doesn't contain required parameter \"uid\".\n" +
                            "Given arguments: $map"
                )
            }

            val adId = map["ad_id"]

            if (adId !is String) {
                throw Exception(
                    "Arguments map doesn't contain required parameter \"ad_id\".\n" +
                            "Given arguments: $map"
                )
            }

            val height: Int = map["height"] as Int? ?: 250
            val width: Int = map["width"] as Int? ?: 300
            val size = Size(width, height)

            val rawParams = map["parameters"]
            var parameters: AdParameters? = null

            if (rawParams is Map<*, *>) {
                parameters = AdParametersFactory.fromMap(rawParams)
            }


            return BannerYandexAdViewArguments(
                uid = uid,
                adId = adId,
                size = size,
                parameters = parameters,
            )
        }
    }
}

File: ios/Classes/platform_api/model_factories/BannerAdViewArgumentsFactory.swift
//
//  BannerViewArguments.swift
//  flutter_yandex_ads_plus
//
//  Created by apple on 21.10.2022.
//

import Foundation

class BannerAdViewArgumentsFactory {
    
    public static func fromMap(map: [String: Any?]) -> BannerYandexAdViewArguments {
        let uid = map["uid"]
    
        if (!(uid is String)) {
            PluginLogger.log(
                message: "Argument \"uid\" is not valid"
            )
        }
    
        let adId = map["ad_id"]
    
        if (!(adId is String)) {
            PluginLogger.log(
                message: "Argument \"ad_id\" is not valid"
            )
        }
    
        let height = map["height"]

        if (!(height is Int)) {
            PluginLogger.log(
                message: "Argument \"height\" is not valid"
            )
        }
    
        let width = map["width"]

        if (!(width is Int)) {
            PluginLogger.log(
                message: "Argument \"width\" is not valid"
            )
        }
    
        let rawParameters = map["parameters"]
        var parameters: AdParameters? = nil
        
        if (rawParameters is [String:Any?]) {
            parameters = AdParametersFactory.fromMap(
                map: map["parameters"] as! [String: Any?]
            )
        }

        return BannerYandexAdViewArguments(
            uid: uid as! String,
            adId: adId as! String,
            height: height as! Int,
            width: width as! Int,
            parameters: parameters,
            settings: BannerAdViewSettingsFactory.fromMap(
                map: map["settings"] as! [String: Any?]
            )
        )
    }
    
    class BannerAdViewSettingsFactory {
        public static func fromMap(map: [String: Any?]) -> BannerYandexAdViewSettings {
            return BannerYandexAdViewSettings(
                translatesAutoresizingMaskIntoConstraints:
                    map[
                        "translates_auto_resizing_mask_into_constraints"
                    ] as! Bool
            )
        }
    }
}
